src/java/org/apache/cassandra/auth/CassandraAuthorizer.java

```diff
@@ -92,18 +96,41 @@ public CassandraAuthorizer()
         }
     }
 
-    public void grant(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource grantee)
+    public Set<Permission> grant(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource grantee)
     throws RequestValidationException, RequestExecutionException
     {
-        modifyRolePermissions(permissions, resource, grantee, "+");
-        addLookupEntry(resource, grantee);
+        String roleName = escape(grantee.getRoleName());
+        String resourceName = escape(resource.getName());
+        Set<Permission> grantedPermissions = Sets.newHashSetWithExpectedSize(permissions.size());
+        Set<Permission> nonExistingPermissions = new HashSet<>(permissions);
+
+        nonExistingPermissions.removeAll(getExistingPermissions(roleName, resourceName, permissions));
+        if (!nonExistingPermissions.isEmpty())
+        {
+            modifyRolePermissions(nonExistingPermissions, resource, grantee, "+");
+            addLookupEntry(resource, grantee);
+            grantedPermissions.addAll(nonExistingPermissions);
+        }
+
+        return grantedPermissions;
     }
 
-    public void revoke(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource revokee)
+    public Set<Permission> revoke(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource revokee)
     throws RequestValidationException, RequestExecutionException
     {
-        modifyRolePermissions(permissions, resource, revokee, "-");
-        removeLookupEntry(resource, revokee);
+        String roleName = escape(revokee.getRoleName());
+        String resourceName = escape(resource.getName());
+        Set<Permission> revokedPermissions = Sets.newHashSetWithExpectedSize(permissions.size());
+        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);
+
+        if (!existingPermissions.isEmpty())
+        {
+            modifyRolePermissions(existingPermissions, resource, revokee, "-");
+            removeLookupEntry(resource, revokee);
+            revokedPermissions.addAll(existingPermissions);
+        }
+
+        return revokedPermissions;
```

Comment:
As before, I think that either `revokedPermissions` or `existingPermissions` is redundant, since one is a copy of the other. We can simplify to:
```suggestion
        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);

        if (!existingPermissions.isEmpty())
        {
            modifyRolePermissions(existingPermissions, resource, revokee, "-");
            removeLookupEntry(resource, revokee);
        }

        return existingPermissions;
```